### PR TITLE
Provide a basic variable hierarchy feature

### LINKED ISSUE #5: Add Variable Hierarchy to GitOpsConfig CR
## What happens?

I would like to be able to use a hierarchy of variables for the process template step.

## What were you expecting to happen?

In a GitOpsConfig CR I can specify a hierarchy of variables. The template processor will handle processing the hierarchy of variables.

## Steps to reproduce: 

* N/A

## Any errors, stacktrace, logs? 

N/A

## Environment:

* Runtime version(Java, Go, Python, etc): N/A
* Desktop OS/version: N/A

## Additional comments:

Examples of variable hierarchies that already exist in the wild ...
* https://github.com/gnuthought/openshift-provision-demo#resource-hierarchy
* https://puppet.com/docs/puppet/5.4/hiera_intro.html

### PR DESCRIPTION
**Pull request by Smiley73**
**Description**
This adds a feature to have a variable hierarchy within the same parameterSource repo. Simply providing a `hiera.lst` file will enable this feature. It's based on the already existing `yq merge` and simply adds processing of multiple directories.

**Additional Changes**
- Updates yq to 2.4.1
- Move configuration files for e2e test into folder `test/e2e/configs`
- Correct ocp e2e test to allow `EUNOMIA_REF` and `EUNOMIA_URI` to work
- Clean up helm test case and use the hello-world example as the base
- Add cluster-admin role to helm example, to support creation of namespace

Fixes #5 

**Type of change**
* New feature (non-breaking change which adds functionality)
* This change requires a documentation update

### CODE REVIEW

**Comment by akavel** [comment]
@vinny-sabatini what are your thoughts on this, esp. with regards to #147? Do I understand correctly that this could replace part of #147 as a somewhat more general approach?

File: README.md [hunk 1]
```diff
@@ -121,6 +121,23 @@ If a secret is provided, then it is assumed that the connection to Git requires
 
 If the `uri` is not specified in the `parameterSource` section, then it will default to the `uri` specified under `templateSource`.
 
+### parameterSource processing
+Eunomia uses the [yq command](http://mikefarah.github.io/yq/) to merge all yaml files in the specified folder. You have to be careful, if you have the same variable name in multiple files. Dictionaries will merge, lists will get overwritten. 
+
+#### Variable Hierarchy
+You can provide a file `hiera.lst`, to allow a variable hierarchy. This will allow for example to specify a default value and overwrite it on an environment level if necessary. This will greatly simplify your configuration and allows for deduplication of data, making your operational life a lot easier.
```
Comment (akavel) [inline README.md]: Why `hiera` and not `hierarchy`? is it some kind of standard name? It seems to me that in the project we  tend to keep full names everywhere with surprising consistency (at least I don't recall many abbreviations), which has the nice benefit of making remembering how to write stuff a bit easier.

File: template-processors/base/bin/processParameters.sh [hunk 2]
```diff
@@ -17,32 +17,50 @@
...
+
+for DIR in ${FOLDERS}; do
+  echo "Processing files in ${DIR}"
+  
+  # get the list of yaml files to process
+  YAML_FILES="$(find ${DIR} -name \*.json -o -name \*.yaml -o -name \*.yml  -maxdepth 1)"
+
+  # merge the files
+  goyq merge --inplace ${VALUES_FILE} ${YAML_FILES}
+done
```
Comment (akavel) [inline template-processors/base/bin/processParameters.sh]: [thinking] Maybe we could consider merging the above blocks into a single pipeline more or less like below:

```bash
echo "---" > "$VALUES_FILE"
cat "$CLONED_PARAMETER_GIT_DIR/hiera.lst" |
    sed -e 's/[[:space:]]*#.*//' |  # delete comments
    grep '\S' |   # prune empty lines
    envsubst |
    (echo "Processing parameters from the following folder(s): " >&2; tee /dev/stderr) |  # print debugging info on stderr
    xargs -n1 -I@ find "$CLONED_PARAMETER_GIT_DIR/@" -name \*.json -o -name \*.yaml -o -name \*.yml -maxdepth 1 |
    xargs goyq merge --inplace "$VALUES_FILE"
```

Though OTOH, it has a notable disadvantage that the debugging info we'd be getting with this change seems of lower quality :/

File: examples/hello-world-hiera/hierademo/hiera.lst [hunk 3]
```diff
@@ -0,0 +1,6 @@
+# yaml files get loaded in the order of the directories listed
+../default # this would be the default
+# ${SOMEVAR}
```
[the remark this answers is not in the document]
Comment (Smiley73) [reply examples/hello-world-hiera/hierademo/hiera.lst]: It's setting the stage for using dynamic values. We first have to add the capability of setting environment variables to eunomia, then we can leverage it.

File: README.md [hunk 4]
```diff
@@ -121,6 +121,23 @@ If a secret is provided, then it is assumed that the connection to Git requires
...
+
+The contents of the file are simply relative path names, with the base being `contextDir`.
+
+Example `hiera.lst`:
+```
+../defaults #this is first
+../marketing #this is second
+../development #this is third
+# lastly the directory containing hiera.lst is merged
```
[the remark this answers is not in the document]
Comment (Smiley73) [reply README.md]: I don't understand your ask. Going without what?
If hiera.lst isn't present, then it defaults to the existing behavior and only loads the yaml files from the directory, which I believe is what you were suggesting.

File: template-processors/base/bin/processParameters.sh (hunk 2 above)
Comment (Smiley73) [reply template-processors/base/bin/processParameters.sh]: That's how I had it originally, but decided to use the current approach to make debugging easier.

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (akavel) [reply README.md]: without the line 137, i.e. "lastly the directory containing hiera.lst is merged"; I'd find it less surprising if we'd have to add the current directory explicitly, as either the last one, or the first one (or any other one). In other words:

- a) `hiera.lst` not present: treated as if `hiera.lst` had contents of: `.`
- b) `hiera.lst` present: no magic handling of the directory containing `hiera.lst`; it's either added by hand in the list, or ignored.

File: examples/hello-world-hiera/hierademo/hiera.lst (hunk 3 above)
[the remark this answers is not in the document]
Comment (akavel) [reply examples/hello-world-hiera/hierademo/hiera.lst]: In such case, can we change this to a `# TODO: ${SOMEVAR}`, and remove the envsubst from resourceManager.sh? to make the intention and status of this feature in this PR clear

File: README.md [hunk 5]
```diff
@@ -121,6 +121,23 @@ If a secret is provided, then it is assumed that the connection to Git requires
...
+Eunomia uses the [yq command](http://mikefarah.github.io/yq/) to merge all yaml files in the specified folder. You have to be careful, if you have the same variable name in multiple files. Dictionaries will merge, lists will get overwritten. 
+
+#### Variable Hierarchy
+You can provide a file `hiera.lst`, to allow a variable hierarchy. This will allow for example to specify a default value and overwrite it on an environment level if necessary. This will greatly simplify your configuration and allows for deduplication of data, making your operational life a lot easier.
+
+The contents of the file are simply relative path names, with the base being `contextDir`.
+
+Example `hiera.lst`:
+```
+../defaults #this is first
```
[the remark this answers is not in the document]
Comment (Smiley73) [reply README.md]: Not sure what you mean. Could you please clarify?

File: examples/hello-world-hierarchy/demo/values.yaml [hunk 6]
```diff
@@ -0,0 +1 @@
+appname: "hello-world-hierarchy"
```
Comment (akavel) [inline examples/hello-world-hierarchy/demo/values.yaml]: another way how politics finds its way into tech? 😜

File: examples/hello-world-hierarchy/cr/hello-world-cr.yaml [hunk 7]
```diff
@@ -1,15 +1,16 @@
...
 spec:
   templateSource:
     uri: https://github.com/KohlsTechnology/eunomia
     ref: master
     contextDir: examples/hello-world-helm/templates
   parameterSource:
-    ref: master
-    contextDir: examples/hello-world-hiera/hierademo
+    #ref: master
+    ref: basic_hiera
```
Comment (akavel) [inline examples/hello-world-hierarchy/cr/hello-world-cr.yaml]: Before merging, this needs to be cleaned up back to `master`, yes?

File: template-processors/base/bin/processParameters.sh [hunk 8]
```diff
@@ -21,45 +21,48 @@ echo "Processing Parameters"
...
-    do
+  while read -r DIR ; do
       # remove everything after # to allow comments and nuke all whitespaces
-      DIR="$(echo -e "${DIR%%#*}" | tr -d '[:space:]')"
+      DIR="$(echo "${DIR}" | sed -e 's/[[:space:]]*#.*//')"
 
       if [ -n "${DIR}" ]; then
         # add the folder to the end of the list
-        FOLDERS="${FOLDERS} ${CLONED_PARAMETER_GIT_DIR}/${DIR}"
+        DIR="$(cd ${CLONED_PARAMETER_GIT_DIR}/${DIR} ; pwd)"
```
Comment (akavel) [inline template-processors/base/bin/processParameters.sh]: please put `cd`'s arg in double quotes

File: template-processors/base/bin/processParameters.sh [hunk 9]
```diff
@@ -21,45 +21,48 @@ echo "Processing Parameters"
...
+        FOLDERS="${FOLDERS} ${DIR}"
       fi
-    done
-fi <<< $(envsubst < ${CLONED_PARAMETER_GIT_DIR}/hiera.lst)
-
-# Add the current folder as the last one to process
-FOLDERS="${FOLDERS} ${CLONED_PARAMETER_GIT_DIR}"
-
-echo "Going to process parameters file in the following folder(s): ${FOLDERS}"
+    done <<< $(envsubst < ${CLONED_PARAMETER_GIT_DIR}/hierarchy.lst)
```
Comment (akavel) [inline template-processors/base/bin/processParameters.sh]: Still the same questions as before:

1. Please mention the cool envsubst feature in the readme! we don't have to be shy towards our users with how fancy we are! :] ideally, it might be cool to have it shown in the examples there
2. Please wrap env vars in `"` — including the `$(...)` expression itself
3. Would it still work if we moved the envsubst before the loop, e.g. like this:

        cat "$CLONED_PARAMETER_GIT_DIR/hierarchy.lst" | envsubst | while IFS= read -r DIR; do
            ...
        done

     to make the data flow more in line with the order of reading the script?

File: template-processors/base/bin/processParameters.sh [hunk 10]
```diff
@@ -21,45 +21,48 @@ echo "Processing Parameters"
 
 FOLDERS=""
 
-if [ -e ${CLONED_PARAMETER_GIT_DIR}/hiera.lst ] ; then
+if [ -e "${CLONED_PARAMETER_GIT_DIR}/hierarchy.lst" ] ; then
   echo "Generating hierarchy"
   # a hierarchy list was provided, so lets process it
-  while read -r DIR
-    do
+  while read -r DIR ; do
```
Comment (akavel) [inline template-processors/base/bin/processParameters.sh]: Please add `IFS=` before `read` — it adds extra protection against squashing of whitespaces in the dir names when doing `read -r` — i.e.:

    while IFS= read -r DIR; do

File: template-processors/base/bin/processParameters.sh [hunk 11]
```diff
@@ -21,45 +21,48 @@ echo "Processing Parameters"
...
+  # No hierarchy provided, so lets just use the current folder
+  FOLDERS="${CLONED_PARAMETER_GIT_DIR}"
+fi
 
-VALUES_FILE=${CLONED_PARAMETER_GIT_DIR}/eunomia_values_processed1.yaml
+# process the folders
+if [ -n "${FOLDERS}" ]; then
+  # ensure our base file exists with one document to make yq happy
+  VALUES_FILE="/tmp/eunomia_values_processed1.yaml"
+  echo "---" > ${VALUES_FILE}
```
Comment (akavel) [inline template-processors/base/bin/processParameters.sh]: please wrap env var in double quotes

File: template-processors/base/bin/processParameters.sh [hunk 12]
```diff
@@ -21,45 +21,48 @@ echo "Processing Parameters"
...
+  for DIR in ${FOLDERS}; do
+    echo "Processing files in ${DIR}"
 
-for DIR in ${FOLDERS}; do
-  echo "Processing files in ${DIR}"
-  
-  # get the list of yaml files to process
-  YAML_FILES="$(find ${DIR} -name \*.json -o -name \*.yaml -o -name \*.yml  -maxdepth 1)"
+    # get the list of yaml files to process
+    YAML_FILES="$(find ${DIR} -name \*.json -o -name \*.yaml -o -name \*.yml  -maxdepth 1)"
```
Comment (akavel) [inline template-processors/base/bin/processParameters.sh]: please wrap DIR env var in double quotes

File: template-processors/base/bin/processParameters.sh [hunk 13]
```diff
@@ -21,45 +21,48 @@ echo "Processing Parameters"
...
 
-  # merge the files
-  goyq merge --inplace ${VALUES_FILE} ${YAML_FILES}
-done
+    # merge the files
+    goyq merge -i -x ${VALUES_FILE} ${YAML_FILES}
```
Comment (akavel) [inline template-processors/base/bin/processParameters.sh]: please wrap VALUES_FILE env var in double quotes

File: template-processors/base/bin/processParameters.sh [hunk 14]
```diff
@@ -21,45 +21,48 @@ echo "Processing Parameters"
...
+  echo "ERROR - no folders found for processing"
+  exit 1
+fi
 
 # Replace variables from enviroment
 # This allows determining things like cluster names, regions, etc.
-if [ -e ${VALUES_FILE} ]; then
-  envsubst < ${VALUES_FILE} > $CLONED_PARAMETER_GIT_DIR/eunomia_values_processed.yaml
+if [ -e "${VALUES_FILE}" ]; then
+  envsubst < ${VALUES_FILE} > /tmp/eunomia_values_processed.yaml
```
Comment (akavel) [inline template-processors/base/bin/processParameters.sh]: 1. Please wrap env var in double quotes
2. Why the change from $CLONED_PARAMETER_GIT_DIR to /tmp/ ? the README seems to require processParameters.sh to put its output in $CLONED_PARAMETER_GIT_DIR ?

File: template-processors/jinja/bin/processTemplates.sh [hunk 15]
```diff
@@ -7,10 +7,10 @@ set -o errexit
 ## we assume that in $CLONED_PARAMETER_GIT_DIR there is a parameter file called parameters.yaml
 ## the result is a set of file stored in the $MANIFEST_DIR, with the same name as the templates, but with no extension. 
 
-for file in $CLONED_TEMPLATE_GIT_DIR/*.j2 ; do
+for file in ${CLONED_TEMPLATE_GIT_DIR}/*.j2 ; do
```
Comment (akavel) [inline template-processors/jinja/bin/processTemplates.sh]: Please put double quotes around the env var, e.g. like this:

    for file in "${CLONED_TEMPLATE_GIT_DIR}"/*.j2 ; do

File: README.md [hunk 16]
```diff
@@ -125,18 +125,21 @@ If the `uri` is not specified in the `parameterSource` section, then it will def
...
+../defaults    #this is first ... lowest priority
+../marketing   #this is second
+../development #this is third ... highest priority
 ```
-In this case it will load all yaml files from `../defaults`, then merge it with everything in `../marketing`, merge it with everything in `../development`, and lastly merge it with everything defined in `contextDir`.
+
+In this case it will load all yaml files from `../defaults`, then merge it with everything in `../marketing`, and lastly merges it with everything in `../development`. You can also use the relative path `./`, which means it'll also load the variables defined in `contextDir` directly (same folder that as `hierarchy.lst`).
+
+#### Upcoming features
+Once [issue #4](https://github.com/KohlsTechnology/eunomia/issues/4) is resolved, you will be able to specify variable names to dynamically determine the correct folder. This will allow you to only have one `hierarchy.lst`.
```
Comment (akavel) [inline README.md]: Suggested addition: *"(Technically, it is actually already possible to use environment variables, but without #4, there are just none set that would be of any practical use in `hierarchy.lst`.)"*

File: examples/hello-world-hierarchy/cr/hello-world-cr.yaml (hunk 7 above)
Comment (Smiley73) [reply examples/hello-world-hierarchy/cr/hello-world-cr.yaml]: Oops. Yes that definitely will need to get cleaned up once I'm done.

File: template-processors/base/bin/processParameters.sh (hunk 9 above)
Comment (Smiley73) [reply template-processors/base/bin/processParameters.sh]: 1. Already done that previously. Will enhance a bit. https://github.com/KohlsTechnology/eunomia/tree/basic_hiera#upcoming-features
2. Done
3. Unfortunately that doesn't work, because it'll start it in a sub-shell and I won't be able to modify the FOLDERS variable. I had it that way before

File: template-processors/base/bin/processParameters.sh (hunk 14 above)
Comment (Smiley73) [reply template-processors/base/bin/processParameters.sh]: 1. done
2. moving it so we don't accidentally process it with the new behavior. fixed the readme.

File: examples/hello-world-helm/service_account_runner.yaml [hunk 17]
```diff
@@ -15,3 +16,17 @@ subjects:
...
+  name: eunomia-demo-runner-cluster-admin
+subjects:
+- kind: ServiceAccount
+  name: eunomia-runner
+  namespace: eunomia-hello-world-demo
+roleRef:
+  kind: ClusterRole
+  name: cluster-admin
+  apiGroup: rbac.authorization.k8s.io
+---
```
Comment (akavel) [inline examples/hello-world-helm/service_account_runner.yaml]: What's the reason for adding this? can you please try to explain in the commit message? I really don't have a slightest idea :( things seemed to be working before, it would really help me understand this code if I knew what's the intention behind it (a.k.a. "why?")

**Comment by seanmalloy** [comment]
Looks good. I cleaned up the commit history a bit.
